compiler/ast/ast.go

```diff
@@ -214,42 +318,56 @@ func (*RecordExpr) ExprAST() {}
 func (*ArrayExpr) ExprAST()  {}
 func (*SetExpr) ExprAST()    {}
 func (*MapExpr) ExprAST()    {}
-
-func (*OverExpr) ExprAST() {}
-
-func (*SQLExpr) ExprAST() {}
+func (*OverExpr) ExprAST()   {}
+func (*SQLExpr) ExprAST()    {}
 
 type ConstDecl struct {
-	Kind string `json:"kind" unpack:""`
-	Name string `json:"name"`
-	Expr Expr   `json:"expr"`
+	Kind       string `json:"kind" unpack:""`
+	KeywordPos int    `json:"keyword_pos"`
+	Name       string `json:"name"`
+	Expr       Expr   `json:"expr"`
 }
 
 type FuncDecl struct {
-	Kind   string   `json:"kind" unpack:""`
-	Name   string   `json:"name"`
-	Params []string `json:"params"`
-	Expr   Expr     `json:"expr"`
+	Kind       string   `json:"kind" unpack:""`
+	KeywordPos int      `json:"keyword_pos"`
+	Name       string   `json:"name"`
+	Params     []string `json:"params"`
+	Expr       Expr     `json:"expr"`
+	Rparen     int      `json:"rparen"`
 }
 
 type OpDecl struct {
-	Kind   string   `json:"kind" unpack:""`
-	Name   string   `json:"name"`
-	Params []string `json:"params"`
-	Body   Seq      `json:"body"`
+	Kind       string   `json:"kind" unpack:""`
+	KeywordPos int      `json:"keyword_pos"`
+	Name       string   `json:"name"`
+	Params     []string `json:"params"`
+	Body       Seq      `json:"body"`
+	Rparen     int      `json:"rparen"`
 }
 
 type TypeDecl struct {
-	Kind string      `json:"kind" unpack:""`
-	Name string      `json:"name"`
-	Type astzed.Type `json:"type"`
+	Kind       string      `json:"kind" unpack:""`
+	KeywordPos int         `json:"keyword_pos"`
+	Name       string      `json:"name"`
+	Type       astzed.Type `json:"type"`
 }
 
 func (*ConstDecl) DeclAST() {}
 func (*FuncDecl) DeclAST()  {}
 func (*OpDecl) DeclAST()    {}
 func (*TypeDecl) DeclAST()  {}
 
+func (x *ConstDecl) Pos() int { return x.KeywordPos }
+func (x *FuncDecl) Pos() int  { return x.KeywordPos }
+func (x *OpDecl) Pos() int    { return x.KeywordPos }
+func (x *TypeDecl) Pos() int  { return x.KeywordPos }
+
+func (x *ConstDecl) End() int { return x.Expr.End() }
+func (x *FuncDecl) End() int  { return x.Rparen }
+func (x *OpDecl) End() int    { return x.Rparen }
+func (x *TypeDecl) End() int  { return x.Type.End() }
```

Comment:
Nits: Put these after their type declarations. And replace x with the first letter of the receiver name.